implement memory profiling

javierluraschi opened the pull request:
Implement memory profiling.
- [x] Profile with memory enabled
- [x] Parse memory in `R` and visualize in `D3`
- [x] Add new memory columns
- [x] Add memory entry to `infobox` and status bar totals
- [x] Minor visual improvements (rounded percentage corners, color adjustments to borders and percentage tooltips)

<img width="1440" alt="screen shot 2016-03-24 at 11 01 40 pm" src="https://cloud.githubusercontent.com/assets/3478847/14039029/ccae6fae-f214-11e5-8387-ad2ec2e118aa.png">

wch commented:
Is d3.js necessary? And similarly, was d3.layout.min.js needed before? I see it was added with treemaps, in f9ff64a9.

javierluraschi commented:
I thought d3.layout it was needed but it's not since newer versions of d3 contain the functionality I need. D3.min is not needed but has helped me debug by swapping the reference in the yaml file, thought it would be nice to include this for other devs, but I can remove it if we think this is unnecessary.

wch commented:
Ah, OK, that makes sense. Another thing you could do is copy d3.js over d3.min.js.

javierluraschi commented:
I like that! Usually we minify files to reduce bandwidth, which is not a big concern here since the packages are downloaded once. Will change.

wch commented on inst/htmlwidgets/lib/profvis/profvis.js:
```diff
@@ -1415,7 +1455,7 @@ profvis = (function() {
           // Size and position the panels
           $panel1.outerHeight($splitBar.position().top - splitBarGap.top -
-                             margin.top);
+                             margin.top - $statusBar.innerHeight());
```
Does the positioning work for a vertical split?

I'm not sure if this was apparent, but the strategy I originally used for placement of the panels is this:

* Do a fixed layout with profvis.css.
* Call `initResizing`. This function takes the original positioning and moves/resizes components to a 50-50 proportion. It preserves the dimensions and spacing of various things.

So if you comment out the call to `initResizing()`, you can position all the elements with CSS. The layout should look roughly correct, though not correctly sized for a particular window height/width. Then when you uncomment `initResizing()`, that should position things correctly.

javierluraschi replied:
Good point, fixed.

wch commented:
Oh, I meant you could copy d3.js over d3.min.js in dev, and leave the minified version in the git repository. But, that said, I think you're right that it's OK to use non-minified d3.js. If you do that, then I think you should edit the yaml file to point to d3.js.

javierluraschi commented:
Ah, right!

wch commented on inst/htmlwidgets/lib/profvis/profvis.js:
```diff
@@ -1730,8 +1772,26 @@ profvis = (function() {
+      }
+
+      return sum;
+    }
```
Missing semicolon?

javierluraschi replied:
Gone.

wch commented on inst/htmlwidgets/lib/profvis/profvis.js:
```diff
@@ -1730,8 +1772,26 @@ profvis = (function() {
+
+        if (node.children) {
+          node.children.forEach(function (child) {
+            nodes.push(child);
```
Nice way of traversing the tree!

javierluraschi replied:
I know, fun times. However, I had to remove this traversal and integrate with `consolidateTree`. At the time I didn't realize that there is a consolidation process that gets rid of duplicate call-stacks, so basically the original pr was missing a bunch of incremental data and the memory visualizations were not accurate. I spent a good part of the day validating cases and running the raw `Rprof` data through R (link-by-line) to validate the memory allocations are correct. So far so good.

wch commented:
General question: When memory profiling is turned on, the profiler includes memory data for each sample. How is that information associated with each line of code? I ask because there can be multiple code lines referred to in each sample. Does it only associate the memory use with the function calls at the top of each stack?


For example, when I run:

```R
library(ggplot2)
Rprof(memory.profiling = T, gc.profiling = T, line.profiling = T)
ggplot(mtcars, aes(wt, mpg)) + geom_point()
Rprof(NULL)
```

The output looks like:

```
memory profiling: GC profiling: line profiling: sample.interval=20000
#File 1: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/plot.r
:521894:3661735:33521824:487:"<GC>" "lazyLoadDBfetch" "gpar" "is.gpar" "valid.viewport" "viewport" "<Anonymous>" "grid.newpage" 1#140 "print.ggplot" "<Anonymous>" 
:521894:3661735:33521824:0:"<GC>" "lazyLoadDBfetch" "gpar" "is.gpar" "valid.viewport" "viewport" "<Anonymous>" "grid.newpage" 1#140 "print.ggplot" "<Anonymous>" 
:514375:1523420:30710176:0:"<GC>" "lazyLoadDBfetch" "gpar" "is.gpar" "valid.viewport" "viewport" "<Anonymous>" "grid.newpage" 1#140 "print.ggplot" "<Anonymous>" 
#File 2: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/plot-build.r
:519661:1563455:31590104:254:"lazyLoadDBfetch" 2#42 "ggplot_build" 1#150 "print.ggplot" "<Anonymous>" 
#File 3: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/ggproto.r
:537733:1644898:36289344:1181:3#72 "fetch_ggproto" 3#92 "$.ggproto" 2#80 "$" 2#80 "ggplot_build" 1#150 "print.ggplot" "<Anonymous>" 
#File 4: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/theme.r
:554191:1676481:40986288:1014:"lazyLoadDBfetch" "<Anonymous>" "lazyLoadDBfetch" "defaults" 4#364 "plot_theme" 2#142 "ggplot_gtable" 1#152 "print.ggplot" "<Anonymous>" 
#File 5: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/margins.R
#File 6: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/theme-elements.r
#File 7: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/utilities-grid.r
#File 8: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/guides-axis.r
#File 9: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/coord-.r
#File 10: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/facet-null.r
#File 11: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/facet-.r
:577774:1757231:45220112:1952:"$" "setChildren" "gTree" 5#98 "titleGrob" 6#182 "element_grob.element_text" 6#136 "element_grob" "grobName" 7#12 "ggname" 6#117 "element_render" 8#43 "guide_axis" 9#53 "f" 3#119 "<Anonymous>" 10#56 "facet_render.null" 11#35 "facet_render" 2#147 "ggplot_gtable" 1#152 "print.ggplot" "<Anonymous>" 
:612888:1774597:52086944:4375:"gtable_add_grob" 2#179 "ggplot_gtable" 1#152 "print.ggplot" "<Anonymous>" 
:548849:1559709:34786360:1652:"grid.Call" "widthDetails.text" "widthDetails" "<Anonymous>" "grid.Call.graphics" "push.vp.viewport" "push.vp" "push.vp.vpTree" "FUN" "lapply" "pushViewport" "pushgrobvp.viewport" "pushgrobvp" "pushvpgp" "preDraw.gTree" "<Anonymous>" "grid.Call.graphics" "push.vp.viewport" "FUN" "lapply" "pushViewport" "pushgrobvp.viewport" "pushgrobvp" "pushvpgp" "preDraw.gTree" "preDraw" "drawGTree" "recordGraphics" "grid.draw.gTree" "grid.draw" 1#154 "print.ggplot" "<Anonymous>" 
:575097:1574768:38225488:2226:"match" "numnotnull" "validGP" "gpar" "is.gpar" "valid.viewport" "viewport" "<Anonymous>" "mapply" "makeContent.gtable" "makeContent" "drawGTree" "recordGraphics" "grid.draw.gTree" "grid.draw" 1#154 "print.ggplot" "<Anonymous>" 
#File 12: /tmp/RtmpXnZbaT/R.INSTALL1d9e40be0a5d/ggplot2/R/grob-absolute.r
:617908:1591088:43502984:3288:"pushedvp" "grid.Call.graphics" "push.vp.viewport" "push.vp" "push.vp.vpTree" "FUN" "lapply" "pushViewport" "pushgrobvp.viewport" "pushgrobvp" "pushvpgp" "preDraw.gTree" "<Anonymous>" "grid.Call.graphics" "push.vp.viewport" "FUN" "lapply" "pushViewport" "pushgrobvp.viewport" "pushgrobvp" "pushvpgp" "preDraw.gTree" "preDraw" "drawGTree" "recordGraphics" "grid.draw.gTree" "grid.draw" "drawGTree" "recordGraphics" "grid.draw.gTree" "NextMethod" 12#46 "grid.draw.absoluteGrob" "grid.draw" "drawGTree" "recordGraphics" "grid.draw.gTree" "grid.draw" 1#154 "print.ggplot" "<Anonymous>" 
```

hadley commented:
How can memory usage decrease by more than 100%?

javierluraschi commented:
Regarding the memory association in the `Rprof` file... it's actually very similar to the way time profiling works: On each given interval, `r-core` logs the current call stack and for memory, it also logs the total amount of memory used in the process, that's about it. From the total memory at a given point in time, we need to re-construct the allocation per-line. In `R` we calculate the diff between each call stack and then in `javascript` we call `consolidateTree()` to aggregate the increments into the right line. The memory profiler is using most of the code that it's already implemented in `profvis` so that will look very familiar.

javierluraschi commented:
@hadley the original "not ready for merge" pr contained an aggregation bug that was displaying the wrong percentages, this is fixed now. The percentages are based on the peak memory allocation.

However, with this definition, I believe it's still possible to percentages in the `abs(x) > 100` range. For instance:

```
d <- NULL

allocMemory <- function() {
    d <<- vector("list", 10)
}

freeMemory <- function() {
    d <<- NULL
}

allocMemory()
freeMemory()
allocMemory()
freeMemory()
```

In this case, I believe the `vector("list", 10)` line could potentially have a -200% since the peak allocation is 10 bytes but gets freed twice (from two different memory blocks) in that single line.

javierluraschi commented:
Ah yes, this happens to be the case... I was able to get this behavior by forcing garbage collection. I think this makes sense since it's consistent with the way we calculate percentages for execution time (they are based on "total tracked execution time"). However, we could consider changing to "percentage of peak memory allocation at a given line of code", that way it won't ever be more than 100% but seems a bit more convoluted, thoughts welcomed.

<img width="1440" alt="screen shot 2016-03-24 at 11 58 54 pm" src="https://cloud.githubusercontent.com/assets/3478847/14039730/7b30a8e2-f21c-11e5-827b-c630cd04ecd3.png">

jjallaire commented:
Maybe shorten "Memory" to "Mem" so that it lines up a bit better with the
column? It it possible to center both the Time and Mem labels over their
columns?

On Fri, Mar 25, 2016 at 3:03 AM, Javier Luraschi <notifications@github.com>
wrote:

> Ah yes, this happens to be the case... I was able to get this behavior by
> forcing garbage collection. I think this makes sense since it's consistent
> with the way we calculate percentages for execution time (they are based on
> "total tracked execution time"). However, we could consider changing to
> "percentage of peak memory allocation at a given line of code", that way it
> won't ever be more than 100% but seems a bit more convoluted, thoughts
> welcomed.
>
> [image: screen shot 2016-03-24 at 11 58 54 pm]
> <https://cloud.githubusercontent.com/assets/3478847/14039730/7b30a8e2-f21c-11e5-827b-c630cd04ecd3.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/profvis/pull/39#issuecomment-201175067>
>

hadley commented:
@javierluraschi I think that suggests computing and displaying % memory at all is a bad idea. The example is especially confusing because the 286 MB allocation is 100%, but the 286 MB deallocation is 200%. It makes sense compute percentage of time because time is always positive. I don't think it make sense to compute a percentage of memory because it's a mix of positive and negative numbers. (It would be ok if you were showing memory allocation over time, but it doesn't make sense when shown over the code space). It's also not clear what "total" memory means in the bottom-right corner. Is that peak allocation?

Generally, I think we should visually play down memory deallocations, since in most cases they are stochastic and are not very revealing. 

BTW I think you also need to make sure that the raw numbers line up correctly/have consistent number of decimals. And the tab title shouldn't be "treemap", because that has an existing meaning. And maybe it should be the first tab, since it's the most important view?

hadley commented:
You might also want to have an example line that both allocates and frees memory - if you're not already, you should probably be aggregating allocs and frees separately.

javierluraschi commented:
@hadley thanks for the feedback, working on it.

Regarding allocations, I'm actually not aggregating separately allocs and frees, I'm tracking the net-allocation. To me, net-allocation makes the most sense since, while investigating memory problems, I'm less concern on memory-churn per module and more about overall allocation per module. For instance, if there is a module that allocs/deallocs a lot of small memory chunks, I wouldn't be concern if the module frees memory appropiately after the call; however, if the net allocation is positive, I would investigate further even if this is a small percent since it could be an indication of memory being leaked. I think giving a visual indication of the net-allocation is more helpful for these investigations.

Let me know what you think, I might be thinking this from a different perspective. Thanks.

javierluraschi commented:
@hadley yes, the bottom right tracks "peak memory allocation". Is there something you would change here? I think this is super useful since you would know how much memory you need in your system before running our of memory or before the os has to start paging memory to disk.

wch commented:
With the memory allocations, I wonder if it really makes sense to say that "line X allocated this much memory".

For the call stack and GC data that's recorded by the profiler, it means that, at a particular moment in time, it's in this function call, or GC event.

If I understand correctly, the memory profiling records the current memory allocation at each sample. So the bars in the table don't indicate how much memory was allocated *by* a particular line of code, they represent how much memory was allocated *when* a line of code is being run. That could be useful information, but it may be misleading, because it only shows memory allocation *after* the memory has been allocated. Consider this example:

```R
library(profvis)
profvis({
  for (i in 1:20) {
    x <- 1:1e6
    pause(0.05)
  }
})
```

![image](https://cloud.githubusercontent.com/assets/86978/14048422/2392ea68-f27c-11e5-86fc-7a52a2bc3a1e.png)


The memory allocation events happen in the line `x <- 1:1e6`, but the bars in the table are large in the line `pause(0.05)`.

There may be a better way to process or display the information, but it's not obvious to me offhand.

wch commented:
One more thought: In the example that I posted above, you can tell from the flamegraph that it's the `x <- 1:1e6` that triggers the memory allocations, because they're also the samples where GC events occurred. It might make sense to include a column in the table that indicates in which lines GC events happened.

javierluraschi commented:
I like the suggestion of the GC column, maybe a non-intrusive visual like circle, not sure if I would want t a full-column but would be helpful to mark this.

Yes, that's a great example Winston... We could, potentially, scope this down to allocations that happened without transitioning callstacks. In the example above we could track that the memory allocation may have come from the previous line or the current one, and we could apply a different visual to allocations that are not well defined, then we can propagate up the callstack with certainty; for code calling this function, the allocations would happen over several intervals and we would know for certainty that the allocation was within this function. I'll take a look to see how feasible this is...

hadley commented:
The problem is that memory allocation is deterministic, but because of the gc, freeing release is  stochastic. If you combine them together, you can't tell if the line allocating 2 MB, or 20 MB, and you just got lucky and the GC kicked in at the same time. The advantage of aggregating them separately is that the display doesn't take up any more space, and it's clear (as in @wch's suggestion) exactly when GC occurs. 

I think displaying peak memory usage in the summary is correct, but that is not what is evoked by total (which implies sum, rather than max)

hadley commented:
We may also want to think more about motivations for using the memory profiler - it's not clear to me why you do want to worry about memory allocation if you already know where the performance bottlenecks are.

javierluraschi commented:
It's certainly the case that cpu bottlenecks are more common; however, for cases where you run out of memory, figuring out which components are allocating the most memory would certainly help users find ways to optimize code. That said, most users of the profiler might not use the memory profiler as often; therefore, it could make sense to hide memory and enable it as an opt-in option from the profiler settings menu.

javierluraschi commented:
@wch @hadley @jjallaire Following up:

1. **Showing memory percentage (Changed)**: I agree, the label it's confusing; therefore, removed. However, I went a step further and also removed percentage for time. Might not seem as confusing as memory percentage, but had this same conversation with Winston when I started using the profiler. I think we would be better of in both cases if the percent is not displayed.
2. **Playing down deallocations visually (Changed)**: Reduced 50% deallocaiton bar width.
3. **Consistent decimals (Changed)**: Set to one single decimal.
4. **Treemap name (Not-Changed)**: I might be misunderstanding, let me know. However, the current view is the "Flamegraph", this is the actual main view, "Treeview" is a different view. Am I missing something? Alternatevely, what names would you propose?
5. **Aggregate allocs/deallocs separately (Not-Changed)**: I still don't think we get much of aggregating this separately since we can still have lines that perform both (allows and deallocs) and we won't be able to split those lines since only the memory total is logged into the Rprof file.
6. **Missleading allocations (Not-Changed, yet)**: Not sure if it's worth the effort changing this since the fix is not trivial, but I'm really open to making this change. How about we use the profiler as-is and I'll collect feedback on this issue to reconsider this change?
7. **Memory profiler motivations (Changed)**: To mitigate the concern of "how useful the memory profiler really is", I added a hide-memory option. It still shows memory by default, but we could switch this easily based on feedback.

My preference is to merge this current iteration.

<img width="1440" alt="screen shot 2016-03-25 at 1 37 20 pm" src="https://cloud.githubusercontent.com/assets/3478847/14053690/51ac478c-f290-11e5-9d1b-9f2e22056586.png">

wch commented:
(1) I agree that it's probably better just to remove percent for both memory and time. People seem to find it confusing.

(6) A few more thoughts about this issue.
  * I have the feeling that the data from the profiler simply doesn't contain enough information for us to display anything useful in hard cases like the one provided above. In that example, the problem is that memory allocation events are almost completely uncorrelated with time -- and the profiler samples with respect to time, not memory allocations. Because the profiler samples wrt to time, it gives us a pretty good picture of time spent in various parts of code. If the profiler sampled with respect to memory allocations (e.g. one sample for every 10MB of memory allocated), then we could use that data to make a good picture of what parts of code cause memory allocations -- but then we would lose the ability to make a good picture of what parts of the code take up a lot of time. There would, for example, be a lot of samples in the `x <- 1:1e6`, but very few samples in the `pause(0.1)`. So in summary, the time-based nature of the profiler means that it's not very useful for memory profiling.
  * All of that said, it's possible that, in practice, those theoretical concerns don't really matter. The current representation of memory profiling data might be roughly "correct" in 95% of real-world use cases. We would still want to make sure users understand what's the numbers and bars actually mean, though. I think we need to try this out in real profiling situations to get a feel for it.
  * With regard to @hadley's comment about the GC being stochastic, I think that's OK for two reasons. First, the profiler is already somewhat random. Second, the GC events are more likely if you're allocating 20MB of memory instead of 2MB, so if you have enough profiling samples, it'll work out in the end; you'll see more GC events at places where large amounts of memory are being allocated.

javierluraschi commented:
Bumping `profvis` version and reverting `d3.min` for two reasons:
 1. It uses 300KB vs 150KB minified. I wasn't expecting D3 to be this large, my bad.
 2. Hit a "d3 not defined" issue only when saving the `htmlwidget` with `selfcontained = TRUE`. If this rings any bells might be worth opening a bug to investigate further since I wouldn't expect this to break while changing toggling the value of `selfcontained`. Could be related to the fact that non-minified version takes +500KB in `base64` form and loaing behavior could be changing per browser, reproed using Chrome.

hadley commented:
Looking good!  I don't feel super strongly about not combining allocs & releases, but IMO it exposes more useful information for very little extra work.

Totally agree with removing percent columns for both. I'm sure I tried and failed to persuade Winston to do this too ;)

javierluraschi commented:
Merging. Happy to iterate on feedback once we obvious improvements become apparent.
